### PR TITLE
add $schema for luis build and qnamaker build dialog files

### LINKED ISSUE #817: luis:build does not include $schema in generated .dialog files
## Versions
@microsoft/botframework-cli/4.10.0-preview.133127 win32-x64 node-v12.16.1

## Describe the bug
A .dialog file should always have a $schema in order to be valid.  With --dialog we generate a .dialog file without a schema.  

## Expected behavior
We should add a --schema switch that if included will set $schema to the value.

[bug]

### PR DESCRIPTION
**Pull request by feich-ms**
fix #817

### CODE REVIEW

**Comment by vishwacsena** [comment]
@chrimc62 can you please review?

File: packages/lu/src/parser/lubuild/cross-trained-recognizer.ts [hunk 1]
```diff
@@ -6,18 +6,24 @@
...
   }
 
   save(): string {
-    let output = {
+    let output: any = {
       $kind: 'Microsoft.CrossTrainedRecognizerSet',
       recognizers: this.recognizers
     }
 
+    if (this.$schema) {
```
Comment (chrimc62) [inline packages/lu/src/parser/lubuild/cross-trained-recognizer.ts:23]: > if (this.$schema) { [](start = 3, length = 20)

Won't this always be true by constructor?

File: packages/lu/src/parser/lubuild/multi-language-recognizer.ts [hunk 2]
```diff
@@ -6,18 +6,24 @@
...
 
   save(): string {
-    let output = {
+    let output: any = {
       $kind: 'Microsoft.MultiLanguageRecognizer',
       recognizers: this.recognizers
     }
 
+    if (this.$schema) {
+      output = {$schema: this.$schema, ...output}
```
Comment (chrimc62) [inline packages/lu/src/parser/lubuild/multi-language-recognizer.ts:24]: Always true

File: packages/lu/src/parser/lubuild/recognizer.ts [hunk 3]
```diff
@@ -7,45 +7,51 @@ import {Settings} from './settings'
...
   save(): string {
-    let output = {
+    let output: any = {
       $kind: 'Microsoft.LuisRecognizer',
       applicationId: this.applicationId,
       endpoint: this.endpoint,
       endpointKey: this.endpointKey
     }
 
+    if (this.$schema) {
```
Comment (chrimc62) [inline packages/lu/src/parser/lubuild/recognizer.ts:51]: Always true.

File: packages/lu/src/parser/qnabuild/cross-trained-recognizer.ts [hunk 4]
```diff
@@ -6,18 +6,24 @@
...
 
   save(): string {
-    let output = {
+    let output: any = {
       $kind: 'Microsoft.CrossTrainedRecognizerSet',
       recognizers: this.recognizers
     }
 
+    if (this.$schema) {
+      output = {$schema: this.$schema, ...output}
```
Comment (chrimc62) [inline packages/lu/src/parser/qnabuild/cross-trained-recognizer.ts:24]: Either remove or allow schema to be undefined.

File: packages/lu/src/parser/qnabuild/multi-language-recognizer.ts [hunk 5]
```diff
@@ -6,18 +6,24 @@
...
   }
 
   save(): string {
-    let output = {
+    let output: any = {
       $kind: 'Microsoft.MultiLanguageRecognizer',
       recognizers: this.recognizers
     }
 
+    if (this.$schema) {
```
Comment (chrimc62) [inline packages/lu/src/parser/qnabuild/multi-language-recognizer.ts:23]: Either remove or allow schema to be undefined.

File: packages/lu/src/parser/qnabuild/recognizer.ts [hunk 6]
```diff
@@ -7,47 +7,52 @@ import {Settings} from './settings'
...
-    let output = {
+    let output: any = {
       $kind: 'Microsoft.QnAMakerRecognizer',
       id: this.id,
       knowledgeBaseId: this.knowledgeBaseId,
       hostname: this.hostname,
       endpointKey: this.endpointKey
     }
 
+    if (this.$schema) {
```
Comment (chrimc62) [inline packages/lu/src/parser/qnabuild/recognizer.ts:52]: > if (this.$schema) { [](start = 3, length = 20)

Either remove or allow schema to be undefined.

File: packages/luis/README.md [hunk 7]
```diff
@@ -368,6 +368,8 @@ OPTIONS
   --suffix=suffix                  Environment name as a suffix identifier to include in LUIS app name. Defaults to
                                    current logged in user alias
 
+  --schema=schema                  Schema value set for dialog files
```
Comment (chrimc62) [inline packages/luis/README.md:371]: >Schema value set for dialog files [](start = 35, length = 33)

Defines $schema for generated .dialog files

File: packages/luis/src/commands/luis/build.ts [hunk 8]
```diff
@@ -40,7 +40,8 @@ export default class LuisBuild extends Command {
     force: flags.boolean({char: 'f', description: 'If --out flag is provided, overwrites relevant dialog file', default: false}),
     luConfig: flags.string({description: 'Path to config for lu build which can contain switches for arguments'}),
     deleteOldVersion: flags.boolean({description: 'Delete old version of LUIS application after building new one.'}),
-    log: flags.boolean({description: 'write out log messages to console', default: false}),
+    log: flags.boolean({description: 'Write out log messages to console', default: false}),
+    schema: flags.string({description: 'Schema value set for dialog files'})
```
Comment (chrimc62) [inline packages/luis/src/commands/luis/build.ts:44]: >Schema value set for dialog files [](start = 40, length = 33)

Defines $schema for .dialog files

File: packages/qnamaker/README.md [hunk 9]
```diff
@@ -148,6 +148,8 @@ OPTIONS
   --suffix=suffix                        Environment name as a suffix identifier to include in qnamaker kb name.
                                          Defaults to current logged in user alias
 
+  --schema=schema                        Schema value set for dialog files
```
Comment (chrimc62) [inline packages/qnamaker/README.md:151]: >Schema value set for dialog files [](start = 41, length = 33)

Define $schema for generated .dialog files

File: packages/qnamaker/src/commands/qnamaker/build.ts [hunk 10]
```diff
@@ -41,6 +41,7 @@ export default class QnamakerBuild extends Command {
     force: flags.boolean({char: 'f', description: 'If --out flag is provided, overwrites relevant dialog file', default: false}),
     qnaConfig: flags.string({description: 'Path to config for qna build which can contain switches for arguments'}),
     log: flags.boolean({description: 'write out log messages to console', default: false}),
+    schema: flags.string({description: 'Schema value set for dialog files'})
```
Comment (chrimc62) [inline packages/qnamaker/src/commands/qnamaker/build.ts:44]: >Schema value set for dialog files [](start = 40, length = 33)

Update comment

File: packages/qnamaker/src/commands/qnamaker/build.ts [hunk 11]
```diff
@@ -41,6 +41,7 @@ export default class QnamakerBuild extends Command {
     force: flags.boolean({char: 'f', description: 'If --out flag is provided, overwrites relevant dialog file', default: false}),
     qnaConfig: flags.string({description: 'Path to config for qna build which can contain switches for arguments'}),
     log: flags.boolean({description: 'write out log messages to console', default: false}),
```
Comment (chrimc62) [inline packages/qnamaker/src/commands/qnamaker/build.ts:43]: >write [](start = 38, length = 5)

Initial cap.

**Comment by chrimc62** [comment]
Awesome to see this!  Minor comments on typing tests and help.

---
In reply to: [640642751](https://github.com/microsoft/botframework-cli/pull/844#issuecomment-640642751) [](ancestors = 640642751)

File: packages/lu/src/parser/lubuild/cross-trained-recognizer.ts (hunk 1 above)
Comment (feich-ms) [reply packages/lu/src/parser/lubuild/cross-trained-recognizer.ts:23]: resolved. allow schema to be undefined.

File: packages/lu/src/parser/lubuild/multi-language-recognizer.ts (hunk 2 above)
Comment (feich-ms) [reply packages/lu/src/parser/lubuild/multi-language-recognizer.ts:24]: resolved. allow schema to be underfined.

File: packages/lu/src/parser/lubuild/recognizer.ts (hunk 3 above)
Comment (feich-ms) [reply packages/lu/src/parser/lubuild/recognizer.ts:51]: resolved.

File: packages/lu/src/parser/qnabuild/multi-language-recognizer.ts (hunk 5 above)
Comment (feich-ms) [reply packages/lu/src/parser/qnabuild/multi-language-recognizer.ts:23]: resolved.

File: packages/lu/src/parser/qnabuild/cross-trained-recognizer.ts (hunk 4 above)
Comment (feich-ms) [reply packages/lu/src/parser/qnabuild/cross-trained-recognizer.ts:24]: resolved.

File: packages/lu/src/parser/qnabuild/recognizer.ts (hunk 6 above)
Comment (feich-ms) [reply packages/lu/src/parser/qnabuild/recognizer.ts:52]: resolved.

File: packages/luis/README.md (hunk 7 above)
Comment (feich-ms) [reply packages/luis/README.md]: resolved.

File: packages/luis/src/commands/luis/build.ts (hunk 8 above)
Comment (feich-ms) [reply packages/luis/src/commands/luis/build.ts]: resolved.

File: packages/qnamaker/README.md (hunk 9 above)
Comment (feich-ms) [reply packages/qnamaker/README.md]: resolved.

File: packages/qnamaker/src/commands/qnamaker/build.ts (hunk 10 above)
Comment (feich-ms) [reply packages/qnamaker/src/commands/qnamaker/build.ts]: resolved.

File: packages/qnamaker/src/commands/qnamaker/build.ts (hunk 11 above)
Comment (feich-ms) [reply packages/qnamaker/src/commands/qnamaker/build.ts]: resolved.

**Comment by feich-ms** [comment]
Comments resolved.

**Pull request merged**